File: packages/php/src/EcollectClient.php
<?php

declare(strict_types=1);

namespace Ecollect;

use Ecollect\Exceptions\InvalidConfigException;
use Ecollect\Modules\CustomersModule;
use Ecollect\Modules\PaymentLinksModule;
use Ecollect\Modules\PaymentsModule;
use Ecollect\Modules\PaymentSystemsModule;
use Ecollect\Modules\ReconciliationModule;
use Ecollect\Modules\SessionModule;
use Ecollect\Modules\TokensModule;
use Ecollect\Modules\WebhooksModule;
use Ecollect\Utils\HttpClient;

/**
 * Main entry point for the ecollect PHP SDK.
 *
 * Usage:
 * ```php
 * use Ecollect\EcollectClient;
 *
 * $client = new EcollectClient([
 *     'api_key'     => 'your-api-key',
 *     'ety_code'    => 123,
 *     'environment' => 'test',
 *     'srv_code'    => 456, // optional default
 * ]);
 *
 * $token   = $client->session->create();
 * $payment = $client->payments->process($paymentIntent);
 * ```
 */
class EcollectClient
{
    /** @var SessionModule */
    public $session;

    /** @var PaymentsModule */
    public $payments;

    /** @var TokensModule */
    public $tokens;

    /** @var WebhooksModule */
    public $webhooks;

    /** @var ReconciliationModule */
    public $reconciliation;

    /** @var CustomersModule */
    public $customers;

    /** @var PaymentSystemsModule */
    public $paymentSystems;

    /** @var PaymentLinksModule */
    public $paymentLinks;

    /** @var Config */
    private $config;

    /**
     * @param  array<string, mixed> $options
     *   - api_key      (string)  Required. Private API key
     *   - ety_code     (int)     Required. Entity/Merchant code
     *   - environment  (string)  Required. 'test' or 'prod'
     *   - srv_code     (int)     Optional. Default service code
     *   - max_retries  (int)     Optional. Default: 3
     *   - initial_backoff_ms (int) Optional. Default: 2000
     *   - log_level    (string)  Optional. 'debug'|'info'|'warn'|'error'. Default: 'info'
     * @throws InvalidConfigException
     */
    public function __construct(array $options)
    {
        $this->validateOptions($options);

        $this->config = new Config($options);

        $http = new HttpClient(
            $this->config->maxRetries,
            $this->config->initialBackoffMs,
            $this->config->logLevel
        );

        $this->session         = new SessionModule($this->config, $http);
        $this->payments        = new PaymentsModule($this->config, $http, $this->session);
        $this->tokens          = new TokensModule($this->config, $http, $this->session);
        $this->webhooks        = new WebhooksModule($this->config, $http, $this->session);
        $this->reconciliation  = new ReconciliationModule($this->config, $http, $this->session);
        $this->customers       = new CustomersModule($this->config, $http, $this->session);
        $this->paymentSystems  = new PaymentSystemsModule($this->config, $http, $this->session);
        $this->paymentLinks    = new PaymentLinksModule($this->payments);
    }

    /**
     * @param  array<string, mixed> $options
     * @throws InvalidConfigException
     */
    private function validateOptions(array $options): void
    {
        if (empty($options['api_key']) || trim((string)$options['api_key']) === '') {
            throw new InvalidConfigException('api_key is required');
        }

        if (empty($options['ety_code']) || (int)$options['ety_code'] <= 0) {
            throw new InvalidConfigException('ety_code must be a positive integer');
        }

        $env = $options['environment'] ?? '';
        if ($env !== 'test' && $env !== 'prod') {
            throw new InvalidConfigException('environment must be "test" or "prod"');
        }
    }
}

File: packages/kotlin/src/main/kotlin/com/ecollect/sdk/EcollectClient.kt
package com.ecollect.sdk

import com.ecollect.sdk.exceptions.InvalidConfigException
import com.ecollect.sdk.modules.*
import com.ecollect.sdk.utils.HttpClient
import okhttp3.OkHttpClient
import java.util.concurrent.TimeUnit

/**
 * Main entry point for the ecollect SDK.
 *
 * ## Server-side usage (with ApiKey):
 * ```kotlin
 * val client = EcollectClient.Builder()
 *     .apiKey("your-private-api-key")
 *     .etyCode(123)
 *     .environment(Environment.TEST)
 *     .srvCode(456)
 *     .build()
 *
 * val payment = client.payments.process(paymentIntent)
 * ```
 *
 * ## Mobile usage (with SessionToken — ApiKey NEVER on device):
 * ```kotlin
 * // Your backend obtains sessionToken using ApiKey, then passes it to the app
 * val client = EcollectClient.fromSessionToken(
 *     sessionToken = "token-from-backend",
 *     etyCode = 123,
 *     environment = Environment.TEST
 * )
 * ```
 */
class EcollectClient private constructor(
    private val config: Config,
    private val httpClient: HttpClient
) {
    val session: SessionModule = SessionModule(config, httpClient)
    val payments: PaymentsModule = PaymentsModule(config, httpClient, session)
    val tokens: TokensModule = TokensModule(config, httpClient, session)
    val webhooks: WebhooksModule = WebhooksModule(config, httpClient, session)
    val reconciliation: ReconciliationModule = ReconciliationModule(config, httpClient, session)
    val customers: CustomersModule = CustomersModule(config, httpClient, session)
    val paymentSystems: PaymentSystemsModule = PaymentSystemsModule(config, httpClient, session)
    val paymentLinks: PaymentLinksModule = PaymentLinksModule(config, payments, session)

    class Builder {
        private var apiKey: String? = null
        private var sessionToken: String? = null
        private var etyCode: Int? = null
        private var srvCode: Int? = null
        private var environment: Environment = Environment.TEST
        private var okHttpClient: OkHttpClient? = null

        fun apiKey(apiKey: String) = apply { this.apiKey = apiKey }
        fun sessionToken(sessionToken: String) = apply { this.sessionToken = sessionToken }
        fun etyCode(etyCode: Int) = apply { this.etyCode = etyCode }
        fun srvCode(srvCode: Int) = apply { this.srvCode = srvCode }
        fun environment(environment: Environment) = apply { this.environment = environment }
        fun okHttpClient(client: OkHttpClient) = apply { this.okHttpClient = client }

        fun build(): EcollectClient {
            val code = etyCode
                ?: throw InvalidConfigException("etyCode is required")

            if (apiKey == null && sessionToken == null) {
                throw InvalidConfigException(
                    "Either apiKey (server-side) or sessionToken (mobile) must be provided"
                )
            }

            val config = Config(
                etyCode = code,
                apiKey = apiKey,
                sessionToken = sessionToken,
                srvCode = srvCode,
                environment = environment
            )

            val client = okHttpClient ?: OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .build()

            return EcollectClient(config, HttpClient(client))
        }
    }

    companion object {
        /**
         * Create a client initialized with a SessionToken (for mobile/frontend use).
         * The ApiKey MUST NOT be stored on mobile devices.
         *
         * @param sessionToken Token obtained by your backend from ecollect
         * @param etyCode Your merchant entity code
         * @param environment TEST or PRODUCTION
         * @param srvCode Optional default service code
         */
        fun fromSessionToken(
            sessionToken: String,
            etyCode: Int,
            environment: Environment = Environment.TEST,
            srvCode: Int? = null
        ): EcollectClient {
            return Builder()
                .sessionToken(sessionToken)
                .etyCode(etyCode)
                .environment(environment)
                .apply { srvCode?.let { srvCode(it) } }
                .build()
        }
    }
}
